### ISSUE TITLE
Replaced proot latest version, but it fails to boot file system

### ISSUE THREAD

**Issue by dinosauria123** [corbinlc/gnuroot#33]
I try to boot Raspbian file system via gnuroot, that succeed. (See issue #14)

Then, I try to replace proot to latest version that distributed on the original proot site.

http://static.proot.me/proot-arm

I renamed proot-arm.bin to proot and placed it /data/data/champion.gnuroot/app_install/support/wheezy/.

I launched file system, but it can not boot.

I need to modify boot script, please give me any suggestions ?

**Comment by cedric-vincent** [comment]
> I launched file system, but it can not boot.

What are the outputs?

> I need to modify boot script, please give me any suggestions ?

First, be sure that the new proot binary is marked as executable since
the "x" bit doesn't survive HTTP download.

Then, because Andoid does not provide a world-writable /tmp, you have
to tell PRoot to use a pre-extracted loader:

```
export PROOT_LOADER=path/to/proot-arm-loader
```

where proot-arm-loader can be extracted from the proot-arm binary this
way:

```
bash$ dd if=proot-arm of=proot-arm-loader bs=1 count=$((16#00008638)) skip=$((16#000b4dfc-16#000ae108+16#09e108))
```

These constants were obtained using ~~sorcery~~^W readelf:

```
bash$ readelf -sW proot-arm | egrep '(loader|data_start)'
   504: 000be2c8     4 OBJECT  LOCAL  DEFAULT   22 loader_path.8296
  3869: 000b4dfc     0 NOTYPE  GLOBAL DEFAULT   21 _binary_loader_exe_start
  4483: 000bd434     0 NOTYPE  GLOBAL DEFAULT   21 _binary_loader_exe_end
  4553: 00008638     0 NOTYPE  GLOBAL DEFAULT  ABS _binary_loader_exe_size
  4738: 000ae108     0 NOTYPE  WEAK   DEFAULT   21 data_start
  5079: 000ae108     0 NOTYPE  GLOBAL DEFAULT   21 __data_start
```

Also, keep in mind that GNURoot provides a extended version of PRoot
that emulates hard links, so even if it seems to work with a "vanilla"
version, you migth run into troubles when installing some packages:

```
https://groups.google.com/d/msg/proot_me/SBemgRlXg2Y/AEc8sE82-MsJ
```

Regards,
Cédric.

PS: @pelya already updated PRoot in Debian noroot (https://github.com/pelya/cuntubuntu/commit/521822d31492004cbfb7aa1839eb07434af1c53e), however I don't know if it is released yet.

**Comment by pelya** [comment]
Yes it is released (version 5.0.0), it crashes Free Pascal that worked
before -
apt-get install fp-compiler-2.6.0 crashes with segfault. I wonder if
Gnuroot also suffers from this. I've patched PRoot to use /tmp inside
chroot, but I'm not entirely sure that my patch works as expected.
I didn't know about PROOT_LOADER, I'll try that.

**Comment by dinosauria123** [comment]
Thank you for reply.

When using PRoot ver.5, terminal quickly crach, I can't catch the output.

When using PRoot ver.4.03 it gave below messages 

proot error: can't create temporary directory: No such file or directory
proot error: can't create glue rootfs

but I can use replaced rootfs and X is also working.

**Comment by cedric-vincent** [comment]
> When using PRoot ver.5, terminal quickly crach, I can't catch the
> output.

Then you could try the PROOT_LOADER trick or a version of PRoot from
Debian noroot.

> When using PRoot ver.4.03 it gave below messages
> 
> proot error: can't create temporary directory: No such file or directory
> proot error: can't create glue rootfs

These [non-fatal] errors are fixed by this patch:

```
https://github.com/pelya/cuntubuntu/blob/521822d31492004cbfb7aa1839eb07434af1c53e/proot-android.patch   
```

@pelya, you can simplify this patch if you know a world-writable
directory at build-time:

```
--- a/src/path/temp.c                                                                                   
+++ b/src/path/temp.c                                                                                   
@@ -10,6 +10,9 @@                                                                                       

 #include "cli/note.h"                                                                                  

+#undef P_tmpdir                                                                                        
+#define P_tmpdir "/foo"                                                                                
+                                                                                                       
```

**Comment by pelya** [comment]
The only world-writable directory on Android is /sdcard, and it's mounted
noexec, so you cannot launch executables from there.
There are no writable paths in /data, except for your app private directory.
If you connect through ADB shell, you can write to /data/local/tmp, but
regular apps cannot write there. You can try to create world-writable
sub-directory there using ADB shell, and use it from the app, but it will
probably not work on Lollipop, because of stricter SELinux policies.

**Comment by dinosauria123** [comment]
I could catch error messages when using ver.5 of PRoot.
I hope that something help.
![error](https://cloud.githubusercontent.com/assets/3067670/5291786/f578f824-7b9a-11e4-93b9-3a7dbb45f38b.png)

**Comment by cedric-vincent** [comment]
Hello @dinosauria123, thanks for  the details.

I'm pretty sure this will be solved either by using:
- the PROOT_LOADER trick (cf. https://github.com/corbinlc/gnuroot/issues/33#issuecomment-65198338); _or_
- this patch from Debian noroot (https://github.com/pelya/cuntubuntu/blob/master/proot-android.patch).

Cédric.
